django/contrib/auth/hashers.py

```diff
@@ -416,35 +453,50 @@ def encode(self, password, salt):
         data = bcrypt.hashpw(password, salt)
         return "%s$%s" % (self.algorithm, data.decode('ascii'))
 
-    def verify(self, password, encoded):
-        algorithm, data = encoded.split('$', 1)
+    def decode(self, encoded):
+        algorithm, empty, algostr, work_factor, data = encoded.split('$', 4)
         assert algorithm == self.algorithm
+        salt, checksum = data[:22], data[22:]
+        return {
+            'algorithm': algorithm,
+            'algostr': algostr,
+            'work_factor': work_factor,
+            'salt': salt,
+            'checksum': checksum,
+        }
+
+    def verify(self, password, encoded):
+        decoded_entry = self.decode(encoded)
+        partial = "".join((decoded_entry['salt'],
+                           decoded_entry['checksum']))
+        data = "$".join(('', decoded_entry['algostr'],
+                        decoded_entry['work_factor'],
+                        partial))
         encoded_2 = self.encode(password, data.encode('ascii'))
         return constant_time_compare(encoded, encoded_2)
 
     def safe_summary(self, encoded):
-        algorithm, empty, algostr, work_factor, data = encoded.split('$', 4)
-        assert algorithm == self.algorithm
-        salt, checksum = data[:22], data[22:]
+        decoded_entry = self.decode(encoded)
         return {
-            _('algorithm'): algorithm,
-            _('work factor'): work_factor,
-            _('salt'): mask_hash(salt),
-            _('checksum'): mask_hash(checksum),
+            _('algorithm'): decoded_entry['algorithm'],
+            _('work factor'): decoded_entry['work_factor'],
+            _('salt'): mask_hash(decoded_entry['salt']),
+            _('checksum'): mask_hash(decoded_entry['checksum']),
         }
 
     def must_update(self, encoded):
-        algorithm, empty, algostr, rounds, data = encoded.split('$', 4)
-        return int(rounds) != self.rounds
+        decoded_entry = self.decode(encoded)
+        return int(decoded_entry['work_factor']) != self.rounds
 
     def harden_runtime(self, password, encoded):
-        _, data = encoded.split('$', 1)
-        salt = data[:29]  # Length of the salt in bcrypt.
-        rounds = data.split('$')[2]
+        decoded_entry = self.decode(encoded)
```

Comment:
Same as above, I'd revert to the simpler version. The main idea behind `decode` for me is a consistent interface where it makes sense, we don't have to bend backwards just to use it.